sw/device/silicon_creator/lib/shutdown.c

```diff
@@ -255,6 +259,57 @@ shutdown_error_redact_t shutdown_redact_policy(void) {
   return shutdown_redact_policy_inline(raw_state);
 }
 
+enum {
+  /**
+   * Length of the hexadecimal representation.
+   */
+  kHexStrLen = 8,
+  /**
+   * Total message length.
+   *
+   * This includes 4 character prefix before the hex string '\n' at the end of
+   * the message.
+   */
+  kErrorMsgLen = kHexStrLen + 5,
+  /**
+   * Base address of UART.
+   */
+  kUartBase = TOP_EARLGREY_UART0_BASE_ADDR,
+  /**
+   * UART TX FIFO size.
+   */
+  kUartFifoSize = 32,
+};
+
+/**
+ * Prints a fixed-length (`kErrorMsgLen`) error message.
+ *
+ * The error message is a concatenation of a 4 character `prefix` (encoded as a
+ * 32-bit value), the hexadecimal representation of `val`, and '\n'.
+ *
+ * @param prefix Prefix encoded as a 32-bit value.
+ * @param val Integer to print.
+ */
+static ALWAYS_INLINE void shutdown_print(shutdown_log_prefix_t prefix,
+                                         uint32_t val) {
+  // Print the 4 character `prefix`.
+  abs_mmio_write32(kUartBase + UART_WDATA_REG_OFFSET, prefix);
+  abs_mmio_write32(kUartBase + UART_WDATA_REG_OFFSET, prefix >> 8);
+  abs_mmio_write32(kUartBase + UART_WDATA_REG_OFFSET, prefix >> 16);
+  abs_mmio_write32(kUartBase + UART_WDATA_REG_OFFSET, prefix >> 24);
+
+  // Print the hex representation of `val`.
+  const char kHexTable[16] = "0123456789abcdef";
+  // `kHexStrLen` is laundered so that it is loaded to a register at every
```

Comment:
Is unrolling this loop a problem? That's the only reason I can see laundering being something we care about here, since there's no loop completion check.